linux: Fix icon not being associated with app window

talal opened the pull request:
This commit fixes the app icon not being correctly associated with the app window. For example, the app icon is not correctly shown in the dock (Gnome on Wayland) when Zed is running.

Release Notes:

- N/A

Commit:
fix: icon not being associated with app window

This commit fixes the app icon not being correctly associated with the app window. For example, the app icon is not correctly shown in the dock (Gnome on Wayland) when Zed is running.

talal commented:
@cla-bot check

ReillyBrogan commented:
This is completely incorrect. You are missing the step of substituting these variables in the .desktop with envsubst if this is the issue you are facing (https://github.com/zed-industries/zed/blob/main/docs/src/development/linux.md#notes-for-packaging-zed). 

For instance, this is how we do it when packaging on Solus:
```
    # vars for envsubst
    export APP_CLI="zedit"
    export APP_ICON="zed"
    export APP_ID="dev.zed.Zed"
    export APP_NAME="Zed"
    envsubst < crates/zed/resources/zed.desktop.in > zed.desktop
```

Adjust accordingly based on whether you are building a preview or the stable version.

talal commented:
@ReillyBrogan Sorry about the initial change that I proposed, I messed up during the copy-paste.

I also use `envsubst`. However, the problem is the missing `StartupWMClass` entry in the `desktop.in` file.

I've updated my commit and changed the hardcoded `StartupWMClass=dev.zed.Zed-Preview` to `StartupWMClass=$APP_ID`.

ReillyBrogan commented:
> I've updated my commit and changed the hardcoded `StartupWMClass=dev.zed.Zed-Preview` to `StartupWMClass=$APP_ID`.

Please see my comment here: https://github.com/zed-industries/zed/issues/12707#issuecomment-2168742761

This is a common point of confusion for people working with .desktop files, especially with how GNOME handles this in a non-standard way. `StartupWMClass` is only needed when the X11 ResourceClass is different than the Wayland AppId. Zed sets the same value for both so it is unnecessary. Instead you should be installing the .desktop file as `$APP_ID.desktop` which will make it work for X11 and Wayland on all DEs and WMs.

mikayla-maki commented:
To make sure I'm understanding this discussion

1. @talal observes that our icon is not associated with our process
2. Proposes a fix that uses a _non standard_, but supported by GNOME, mechanism
3. @ReillyBrogan says that the better fix is to change the way our desktop file is named to `$APP_ID.desktop`

Is that correct?

talal commented:
On second thought, I think the solution proposed by @ReillyBrogan is better; use `$APP_ID.desktop` as the filename for the desktop file.

For posterity, in case any other Gnome user comes across this issue: use the [install script](https://github.com/zed-industries/zed/blob/main/script/install.sh) for installing Zed, it will install the desktop file with the correct filename.

mikayla-maki commented:
Since this solution doesn't hurt, and applies specifically to non-bundled zed, I think I' ll just merge it so that the GNOME experience is that much better :)

ReillyBrogan commented:
I don't think it should be merged. The correct solution is to rename the `.desktop` file to `$APP_ID.desktop`. That will solve this issue for all of the following:
- GNOME Wayland and X11 sessions
- Plasma Wayland and X11 sessions
- Every other desktop environment or window manager that implements this aspect of the FreeDesktop standards, which is basically all of them

Whereas having `StartupWMClass` only solves it for:
- GNOME Wayland and X11 sessions

Given that GNOME is the most popular DE and that packagers are often themselves using it doing this basically only masks the real proplem that they could otherwise solve correctly for everyone by renaming the desktop file.

On Solus Linux we do all packaging updates as PRs on our packaging recipe monorepo. We have checklists for those reviewing PRs that directly mention that a packager adding the field manually to a desktop file indicates that the desktop file may be incorrectly named in the first place. The number of instances where the field is the correct solution to a problem is dramatically lower than the number of times the field is used by people when they should have renamed the desktop files or patched the code to use an app_id that corresponds with how they named the desktop file.

`StartupWMClass` is the correct solution when an X11 window has like a splash or loading screen style window that appears before the main application window appears and that disappears once the main window is present. It is the incorrect solution in every single other instance. (Wayland deals with this differently by having a concept of parent and child app_id relationships that loading screen windows can use so that they're correctly associated by the compositor).

Really the best solution here is to revert this change and update the documentation so that it's obvious that the desktop file needs to be named the same as the $APP_ID, with a reference to the file in the code that needs to be patched if the desktop file needs to be named using a different standard (like distro policies).

Honestly, it's looking we probably need an equivalent to a `make install` that correctly names the desktop file and pre-fills it with the correct settings using whatever the app was built with. Otherwise you'll have a bunch of packages on different distros that are wildly different. Like some might not ship shell completions once those are available while others do, leading to an inconsistent experience.

mikayla-maki commented:
@ReillyBrogan Yes, let's improve the documentation for building Zed. We are already assuming that your build process will need to adjust the desktop file, and we do so in our own building step, so we should let others know. I don't think this PR harms anything though, and it might make things slightly better in some narrow edge cases, so I'm going to leave it as is. If you can show that this makes things worse in some way, then I'm happy to revert it. But frankly, I'm not interested in GNOME vs. everyone else conflicts. I just want Zed to work wherever it happens to be.
